### ISSUE TITLE
xIPAddress:Static IP is not reflecting in consistency report 

### ISSUE THREAD

**Issue by rsganta** [PowerShell/xNetworking#277]
Hi ,
I have a scenario to assign statip address, gateway,subnetmask,dnsserveraddresses.  I am able to perform all those using DSC configuration after joining the machine to private domain. However in azure portal-> consistency report check says status as Inprogress for this node and it does not report new IPAddress. instead it shows old IP and configuration mode as blank. Any idea why status says InProgress ( I am expecting Compliant) also why IPAddress and configuration mode was not reflecting. Any help would be greatly appreciated for better understanding and correcting myself.

Node state at report time:

Node name
<nodename>
IP address
<old ip address>
Configuration mode
--

**Comment by rsganta** [comment]
Node name:
node a
ipaddress
old ip address (not new one which was set using configuration)
configuration mode
blank

**Comment by PlagueHO** [comment]
Hi @rsganta - I am wondering if you change the IP Address etc on the Azure VM if it prevents the machine reporting the DSC configuration status. Are you able to remote into the VM and run the following PowerShell commands:
```powershell
Get-DscLocalConfigurationManager
Get-DscConfiguration
```
and post the results?

**Comment by rsganta** [comment]
Hi PlagueHO, I am using OnPremVM and not an azure VM.  My goal is to deploy a vm and rename , assignip,subnetmask,gateway and Dns server addresses. I would like to rename the VM as part of deploy VM .  Once rename is done, I would like to use DSC to perform other actions.  Does this approach work?

**Comment by PlagueHO** [comment]
Hi @rsganta - the part that I'm not sure about is "renaming the VM" - do you mean changing the Compter Name of the virtual machine or change the name of the VM entry in Hyper-V? Both are possible but require a very different approach.

But if I assume that you're wanting to rename the Computer Name of the VM. You can do all of these tasks. I actually do something similar to this in my [LabBuilder](https://github.com/PlagueHO/LabBuilder) project (for building Hyper-V labs using DSC).

What I do when provisioning the VHD/VHDx for the OS on the Hyper-V host is I create and inject a DSC Configuration for the machine that configures the IP Address/Subnet, Default Gateway and DNS Addresses directly onto the VHD disk of the new VM (before first boot). I also inject a `SetupComplete.cmd` file (see https://technet.microsoft.com/en-us/library/cc766314%28v=ws.10%29.aspx?f=255&MSPPError=-2147217396) that calls a custom PowerShell script that will start application of the DSC when it is run. I'll also create an `Unattend.xml` file that configures the computer name for me and sets things like timezones/localization (could be done in DSC but reduces reboots).

You can see this code here (but there is a lot of other stuff going on here as well):
https://github.com/PlagueHO/LabBuilder/blob/dev/LabBuilder/lib/private/vm.ps1#L80

But another great alternative that could solve many of these issues for you is [Lability](https://github.com/VirtualEngine/Lability) worth taking a look as I'm certain many of your questions are addressed by these tools - so you don't even have to worry about it.

**Comment by rsganta** [comment]
Thank you PlagueHO,
You are correct, I want to rename the computer. I also use unattend.xml and copying LCM.ps1 to c directory and running with powershell. It seems , I am looking for the similar solution . Let me take a look at your sample . AS I will be on vacation in the next few weeks and I would keep this thread open for my tracking.

**Comment by PlagueHO** [comment]
Sure @rsganta - I'll leave this open till you get back. Thank you!

**Comment by rsganta** [comment]
Hi PlagueHO,
I am back from vacation. I am going to actively look into this solution this week and update you.
Thank you for your patience.
Ravi

**Comment by PlagueHO** [comment]
No worries @rsganta . I'm about to have a holiday for a few days :grin:. So I'll have a look at anything you feedback here after Sunday.

**Comment by rsganta** [comment]
Thank you.  I was going through email again and found that you have injected ip config,subnet etc into VHD.  In my case, I cannot inject IP into VHD because I would like to build 10 VMs which will have different IP Addresses and also all these 10 VMs should be renamed to different VM names. Irrespective of this, I am still going through the solution to understand if the same can be applied to solve my problem.

**Comment by rsganta** [comment]
Also, Since I am creating this onprem VM using azure runbook  via hybridworker group. below is the sample syntax 
$newVM = New-VM -Name $ServerName -SwitchName $($vSwitch[0].name) -MemoryStartupBytes $memoryStartupBytes  -VHDPath $targetVHDFolder\$($ServerName).vhdx  -BootDevice VHD -Generation 2
After the above line run in runbook, is there any option that we can use powershell remoting and rename the newly created VM ?

**Comment by rsganta** [comment]
I was able to rename the vm by using rename-computer after enabling wmi using netsh advfirewall. I will try to figure out to see how I can make it dynamic using powershell runbook.  This may help my overall approach

**Comment by PlagueHO** [comment]
Are you sure you can inject the basic machine information into the VHX prior to first start? You would typically configure a unique machine name on a machine at first boot servicing by copying a Windows Unattended Answer file into the VM. This would cause the VM to boot up with the appropriate computer name. You can also inject a script into the VM prior to first boot that could contain the IP Address you wanted to set the VM to.

To inject the script and unattend.xml file you can just mount the VHDX file onto a host and copy the files in. I think there are also some DISM commands that can be used for this as well (it has been a while since I worked on this).

So, prior to executing Start-Vm for the first time on each VM (just after executing the New-Vm):
1. Mount the VHDX using `$mountedvhd = Mount-VHD -Path $targetVHDFolder$($ServerName).vhdx`
2. Generate an unattend.xml file that contains the name of the machine (this is easy enough but an Unattend.xml file is pretty large) and copy it into the mounted VHD.
3. Generate a script that will run at first boot (can't remember the name/path this goes) that configures the IP Address of the adapter - the copy it into the mounted VHD.
4. Dismount the VHD `Dismount-VHD`.
5. Boot the VM.

Sorry about the lack of details, but this is how I do this in LabBuilder and although there are definitely other approaches this is fairly easy.

HTH

**Comment by rsganta** [comment]
Thanks Plague. I am still trying to understand how the above steps works. Below is my code . I am creating another harddisk ,setting processor count between create and start of the vm. 

I would like to  add below 2 steps  between create and start of the vm.  I still need to figureout , how I can get machine name (which was created with random name) with new-vm option. Any insights would be great.

enable wmi on newly created vm.
 netsh advfirewall firewall set rule group="windows management instrumentation (wmi)" new enable=yes
Rename VM:
Rename-Computer -ComputerName "NewTestVM5" -NewName "DSC-Share" -LocalCredential "NewTestVM5\administrator" -DomainCredential "NewTestVM5\administrator" -PassThru -Restart -Force


Create vm script:

$newVM = New-VM -Name $ServerName -SwitchName $($vSwitch[0].name) -MemoryStartupBytes $memoryStartupBytes  -VHDPath $targetVHDFolder\$($ServerName).vhdx  -BootDevice VHD -Generation 2

    If ($newVM -ne $null)
    {
        #Step 2a: Set the number of processors for the VM:   
        Write-Output "Set VM Processor Count to $($vmprocessorcount)"
        SET-VMProcessor -VMname $($newVM.name) -count $vmprocessorcount #parameters (If user specified value is greater than 8, use max available vprocessors)
          
        #Rename-Computer -NewName "RaviTest123" -Restart
        #handled through DSC module
        
        Write-Output "Create a virtual hard data disk  with dynamic size() on VM $($newVM.Name)"
        $VM = Get-VM -Name $newVM.Name  # may not be needed
        #$currentVHDPath = $vm[0].HardDrives
        $SCSIController = Get-VMScsiController -VMName $vm.Name
        $AvilableSCSIController = $SCSIController.ControllerNumber
        $newControllerLocation = $SCSIController.ControllerNumber + 1
     
        #Create a vhd on the host and mount it
        New-VHD -Path $targetVHDDataFolder -SizeBytes 250Gb -Dynamic |Mount-VHD -Passthru |Initialize-Disk -PassThru |New-Partition -AssignDriveLetter -UseMaximumSize |Format-Volume -FileSystem NTFS -Confirm:$false -Force
        #Dismount the VHD
        Dismount-VHD -Path $targetVHDDataFolder
        Add-VMHardDiskDrive -VMName $VM.Name -ControllerType SCSI -ControllerNumber $AvilableSCSIController -ControllerLocation $newControllerLocation -Path $targetVHDDataFolder

        #Start the Server
        Write-Output "Starting the VM $($newVM.Name)"
        Start-VM -Name $newVM.Name

**Comment by rsganta** [comment]
Please let me know if there is any efficient way through DSC to rename the VM ,assignIPAddress and join to domain with a single LCM file.

**Comment by PlagueHO** [comment]
Hi @rsganta - sorry about the delay on getting back to you.

The way I've done this in LabBuilder is to add a few steps after the `New-VHD` command in your code above. I'll try and summarize the steps and link to the code in LabBuilder (which you should be able to copy mostly):

1. Mount the boot VHD to a folder on the host (I use a temporary folder as the mount point): https://github.com/PlagueHO/LabBuilder/blob/dev/LabBuilder/lib/private/vhd.ps1#L54

2. Create a folder on the mounted VHD to store the script files: https://github.com/PlagueHO/LabBuilder/blob/dev/LabBuilder/lib/private/vhd.ps1#L226

3. Create (via automation) the following scripts (this is done throughout LabBuilder):
  a. The DSC Configuration for the node.
  b. The LCM configuration for the node.
  c. A CMD script that executes a PowerShell script
  d. the PowerShell script executed by the CMD that will apply the LCM configuration and the apply the DSC Configuration.
Note: I apply an LCM configuration because I'm using MOF encryption - so you have to configure the LCM for that.

4. Copy the above files into the folder created in 2.

5. Create an unattend.xml that will execute the CMD script that will then call a PowerShell script that will apply the LCM configuration.

6. Copy the unattend.xml into the appropriate folder (\Windows\Panther\Unattend.xml) on the VM: https://github.com/PlagueHO/LabBuilder/blob/dev/LabBuilder/lib/private/vhd.ps1#L248

7. Dismount the mount point.

8. Boot the VM.

But to be honest, if you're going to do all this then I'd just say why not use LabBuilder to do it all? This is the exact scenario I created it for: I wanted to be able to create a bunch of VMs on a Hyper-V host and have DSC configure everything on them - without having to manually configure the DSC on each one.

It might take a bit of time to learn to configure your own labs, but it'll probably be less time than writing the code to do DSC/LCM injection it fresh VM's from scratch. I wrote an introduction to it a while ago: https://dscottraynsford.wordpress.com/2016/10/04/easily-create-a-hyper-v-windows-server-2016-ad-nano-server-lab/

It comes with a bunch of preconfigured Labs that you could customize.

**Comment by rsganta** [comment]
Hi @PlagueHO  , 
Thanks for the info. I have created similar solution but different way and would like to add steps here.
Phase 1: Using unattended.xml, created a master .vhd (during creation, I have created folder called DSC and where I put my LCM file) -> Manually done
Phase 2: Created a powershell run book to create A VM using a master vhd created in phase 1.
Result: VM gets created with random machine name WIN-XXXXXXX -> Automated
I want to modify the powershell runbook to create the vm's computer name with the name I want it for let's say ABC.  If I could do this,  Since DSC already have a config with node ABC, it should work

To work around in renaming computer to ABC, I have to do two manual steps (1-Enable WMI on the computer , 2- From VMHost, need to rename the computer to ABC). Then I have to wait for DSC to bring my changes.. If I solve those 2 steps , it would help me to avoid these manual steps.

**Comment by rsganta** [comment]
Hi @PlagueHO ,
Here is my runbook code, As i am trying to see how i can use the lab builder and could not map the correct lines of code for my purpose.  I would greatly appreciate, If you could help in the [TBD] section of the below runbook code 
a) how to find a computername after vm gets created b) how to enable wmi on the newly created vm c) rename the computer.

 Write-Host "Create a VM from image with -MemoryStartupBytes $($memoryStartupBytes)"
    $newVM = New-VM -Name $ServerName -SwitchName $($vSwitch[0].name) -MemoryStartupBytes $memoryStartupBytes  -VHDPath $targetVHDFolder\$($ServerName).vhdx  -BootDevice VHD -Generation 2

    If ($newVM -ne $null)
    {
        #Step 2a: Set the number of processors for the VM:   
        Write-Host "Set VM Processor Count to $($vmprocessorcount)"
        SET-VMProcessor -VMname $($newVM.name) -count $vmprocessorcount #parameters (If user specified value is greater than 8, use max available vprocessors)
        
        [TBD]
		# How to find out computer name after VM gets created ?? If i can find the computer name, e.g. srv01 , it can be used as old name for the rename-computer
		# Before rename-computer , need to enable WMI on the newly creted VM using the below command
		#netsh advfirewall firewall set rule group="windows management instrumentation (wmi)" new enable=yes  # to enable WMI
        #Rename-Computer -ComputerName "Srv01" -NewName "NewSrv01" -LocalCredential Srv01\Admin01 -DomainCredential Domain01\Admin01 -Force -PassThru -Restart   #T0 rename computer
        
        
        Write-Host "Create a virtual hard data disk  with dynamic size() on VM $($newVM.Name)"
        $VM = Get-VM -Name $newVM.Name  # may not be needed
        #$currentVHDPath = $vm[0].HardDrives
        $SCSIController = Get-VMScsiController -VMName $vm.Name
        $AvilableSCSIController = $SCSIController.ControllerNumber
        $newControllerLocation = $SCSIController.ControllerNumber + 1
     
   
        #CreateVHD -VHDPath $targetVHDDataFolder -Size 250GB  #$targetVHDFolder\$($ServerName).vhdx

        $drive = (New-VHD -path $targetVHDDataFolder -SizeBytes 250GB -Dynamic   | `
                Mount-VHD -Passthru |  `
              get-disk -number {$_.DiskNumber} | `
              Initialize-Disk -PartitionStyle MBR -PassThru | `
              New-Partition -UseMaximumSize -AssignDriveLetter:$False -MbrType IFS | `
              Format-Volume -Confirm:$false -FileSystem NTFS -force | `
              get-partition | `
              Add-PartitionAccessPath -AssignDriveLetter -PassThru | `
              get-volume).DriveLetter

         Dismount-VHD $targetVHDDataFolder #$VHDPath
         MountVHD -VHDPath $targetVHDDataFolder
         DismountVHD -VHDPath  $targetVHDDataFolder 
         Add-VMHardDiskDrive -VMName $VM.Name -ControllerType SCSI -ControllerNumber $AvilableSCSIController -ControllerLocation $newControllerLocation -Path $targetVHDDataFolder 
        #Get-Disk -Number 1| Set-Disk 1 -IsOffline $false
 

        #Start the Server
        Write-Host "Starting the VM $($newVM.Name)"
        Start-VM -Name $newVM.Name
        Start-Sleep -s 60

**Comment by rsganta** [comment]
btw, I am using hyper V hosted VM.

**Comment by PlagueHO** [comment]
Hi @rsganta - why don't you set the name of the computer in the Unattend.xml file? That way the guest VM will have the correct name when it first boots?

**Comment by rsganta** [comment]
@PlagueHO  , 
Since I have to create many vms, It may not be efficient to generate vhds using those unattended .xmls.

**Comment by PlagueHO** [comment]
@rsganta - are you wanting to automatically rename the guest computer name with the name of the machine in Hyper-V after the machine has completed it's OS install? This is possible, but it is definitely more complicated. I'm do still think it would just be simpler to create a base VHDx then create differencing disks for each VM and inject a custom unattend.xml file into each one as you create it. This might be a bit of code, but I'd expect no more than renaming the machine from the host after it has completed installation.

But if you really want to go down the route of renaming the machine from within the Host your best bet is to use [PowerShell Direct](https://blogs.technet.microsoft.com/virtualization/2015/05/14/powershell-direct-running-powershell-inside-a-virtual-machine-from-the-hyper-v-host/) to remote into the VM after it has completed installation (that in itself can be difficult to determine) and then rename it:

Using your code:
```powershell
If ($newVM -ne $null)
{
    #Step 2a: Set the number of processors for the VM:   
    Write-Host "Set VM Processor Count to $($vmprocessorcount)"
    SET-VMProcessor -VMname $($newVM.name) -count $vmprocessorcount #parameters (If user specified value is greater than 8, use max available vprocessors)
    
Invoke-Command -VMName $newVM.name -ScriptBlock { Rename-Computer -NewName "NewSrv01" -Force -PassThru -Restart }
```

That should work, but I will reinforce again, you're going to need to ensure the VM has finished completing the installation before trying this process.

HTH!

**Comment by johlju** [comment]
Would it be possible to use xComputer in the https://github.com/PowerShell/xComputerManagement Resource module to rename the computer name?

**Comment by PlagueHO** [comment]
@johlju is quite correct. xComputer can be used to rename a machine. But I think from memory in this case @rsganta wanted to be able to apply the MOF from the VM host - which meant the machine name needed to already be set to a known state. E.g.
```powershell
Start-DscConfiguration -ComputerName 'GuestVMName' ....
```

But one other idea would be to use PS Direct to load the MOF into the guest VM and then apply it from a PS Direct session:
```powershell
$session = New-PSSession -VMName $newVM.Name
Copy-Item -Path 'c:\moffiles\GuestVM1.mof' -DestinationPath 'c:\moffiles\LocalHost.mof' -ToSession $session
Invoke-Command -Session $session -ScriptBlock { Start-DscConfiguration -ComputerName Localhost -Path 'd:\moffiles' -Wait -Verbose -Force }
Remove-PSSession -Session $session
```

That way you don't need to know the machine name - and the DSC Config could use xComputer to rename the computer as @johlju suggested.

**Comment by rsganta** [comment]
thanks @johlju  and @PlagueHO , I tried to use one of the above logics 
Invoke-Command -VMName $newVM.name -ScriptBlock { Rename-Computer -NewName "NewSrv01" -Force -PassThru -Restart }

With that 
1) It is asking username and password. May be we can use poweshell credentials to overcome this.
2) It complaints that the VM did not start.

**Comment by PlagueHO** [comment]
Hi @rsganta - you may have to supply the `-Credential $credential` parameter to the Invoke-Command.

Also, the `Invoke-Command` should have been run after starting the VM with `Start-VM` and then waiting until the VM has completed installing (which may take a while). As I mentioned: it isn't easy to identify if the VM has completed installation - so one thing you can do is keep trying to reconnect until successful, but you'd want to implement some sort of max time to wait. There are other things you can do, but they are more complex and involve creating files from scripts executed in the unattend.xml - and if you're going to do that, why not just rename the computer there.

But again, I'm still not completely sure why you're not using the unattend.xml to do all this. It is what it was designed for.

**Comment by rsganta** [comment]
Thanks @PlagueHO . Trying to understand and exploring multiple options. Since i have already generated powershell runbook and by adding little logic might address the issue. Also i am trying to create 3 hard disks in case of sql. My current run book, i am creating a VHD and attaching VHD to VM in one instance which is good for share point vms. For sql VMs, we need 3 drives C,D & E.

**Comment by rsganta** [comment]
Also, Any other options to retrieve the computername(original vmname) once hyper V gets created the VM ?

**Comment by rsganta** [comment]
@PlagueHO ,
I was going through the option you suggested using setupcomplete.cmd and unattend.xml. Couple of questions for my clarification.  Let me explain the current state where I am in and what else to be done on my end.

Current state:  
1) I am able to create a VM using .vhdx with 3/2 disks based on the nodename. 
2) This VM does not have IP Address and vm name that I want to assign.
3) Created a local directory with dscconfiguration.ps1, which runs on the first reboot( as said above, it does not assign with IP and intended computername)

Expecting state:
1) Disks to be online: Once VM is up and running set the newly added disks(not OS disk) to online. To my knowledge, this can be done in two ways a) If i know the VMname from hyper V host, i can use script block (as suggested above) and execute in in $pssession b)  While creating VHD , add the steps to setupcomplete.cmd  and hence once VM is up and running it will execute the script.
2) IP Address:  Since i am going to create 10 vms with different IP Addresses, how can I assign them?
3) Rename of the computer: Though i want to add this step to setupcomplete.cmd, name varies from node to node..
Also, could you please share your setupcomplete.cmd and .ps1 files ?

**Comment by rsganta** [comment]
@PlagueHO , Would be you be able to provide email and hence i can work with you offline to understand it better and  I will update the thread after i resolve the issue. Please let me know if this is feasible.

**Comment by rsganta** [comment]
ping

**Comment by PlagueHO** [comment]
Hi @rsganta - sorry about the delay. I've just been off for a little bit. I'll look through this and respond fully later on in the week. You can email me on dscottraynsford@outlook.com.

**Comment by rsganta** [comment]
Thanks @PlagueHO

**Comment by PlagueHO** [comment]
Hi @rsganta - Sorry for taking so long to get back to you. But in answer to your 3 questsions:
1. To bring the other disks online I'd go for option a. You could then apply a MOF file directly to the VM using a PowerShell Direct (`invoke-command -VMName VM1 { Start-DscConfiguration ... }`) You could use a DSC Config that uses the xStorage module to bring this disk online and partition/format it.
2. I'd suggest you build 10 DSC configs on the VM Host and compile them to MOF files. Then using a loop, remote to each one using PowerShell Direct and apply the MOF (just like in 1 above).
3. You can set the name of the node using xComputer resource. Again, you could include this in your DSC config - so you'd compile 10 of these and then apply each one using PowerShell Direct.

But basically you could do something simple like this (this code is just a sample):
```powershell
$vms = @(
    @{
        vmname = 'VM1'
        computername = 'mynode1'
        ipaddress = '192.168.1.10'
    },
    @{
        vmname = 'VM2'
        computername = 'mynode2'
        ipaddress = '192.168.1.11'
    }
)


$destpath = 'd:\dscconfigs'
. dsctest.ps1
foreach ($vm in $vms) {
    myawesomevm -OutputPath $destpath -Name $vm.vmname -ComputerName $vm.computername -IpAddress $vm.ipaddress
    $pssession = New-PSSession -VMName $vmname -Credential $credentialtoVM
    Copy-Item -Path "$destpath\$vmname.mof" -Destination c:\localhost.mof -ToSession $pssession
    Invoke-Command -Session $pssession -ScriptBlock { Start-DscConfiguration -Path c:\ -Force -Wait -ComputerName localhost }
}

Configuration myawesomevm
{
    param (
        [System.String] $Name,
        [System.String] $ComputerName,
        [System.String] $IpAddress
    )

    Import-DscResource -ModuleName xNetworking, xComputerManagement, xStorage

    # DSC Config goes here...
}
```

You could do something like the above - but it will require a lot more work than just this.

**Comment by PlagueHO** [comment]
Also, have you checked out Lability (https://github.com/VirtualEngine/Lability) - this might also do everything you need. It really does seem like you might be doing a lot of work you don't need to do (because someone else has already done it for you :grin: )

**Comment by rsganta** [comment]
Thanks @PlagueHO 
Sure. I will check these to see if they can help.

**Comment by rsganta** [comment]
Finally, I have chosen path to proceed to have a ipaddress.ps1 file with servernames and IPAddresses which can be injected with Master VHD. During my VM creation, I also injects servername dynamicall in a .txt file . I have compared the content in .txt file with the data in ipaddress.ps1 . If it matches the servername I am assigning IPAddress and rename the computer. This solved the problem.  

I would like to call out everyone who helped me in this regard and thanks much.These discussions helped me in more ways to find out the solutions.

Thanks @PlagueHO

**Comment by PlagueHO** [comment]
Thanks @rsganta - glad we were able to help and you got this sorted! Awesome stuff! :smile:
